RFC/src/RFC-0201_TariScript.md

```diff
@@ -0,0 +1,410 @@
+
+### Scriptless scripts
+
+[Scriptless script] is a wonderfully elegant technology and inclusion of Tari script does not preclude the use of
+Scriptless script in Tari. However, scriptless scripts are difficult to reason about and development of them are best
+left to experts in cryptographic proofs, leaving the development of Mimblewimble smart contracts in the hands of a very
+select group of people.
+
+However, it is the opinion of the author that there is no reason why Mimblewimble cannot be extended to include
+scripting.
+
+## Tari script - a basic motivation
+
+The essential idea of Tari script is as follows:
+
+Given a standard Tari UTXO, we add _additional restrictions_ on whether that UTXO can be included as a valid input in a
+transaction.
+
+As long as those conditions are suitably committed to, and are not malleable throughout the existence of the UTXO, then
+in general, these conditions are no different to the requirement of having range proofs attached to UTXOs, which require
+that the value of Tari commitments is non-negative.
+
+Note that range proofs can be discarded after a UTXO is spent, since the global security guarantees of Mimblewimble are
+not concerned that every transaction in history was valid from an inflation perspective, but that the net effect of all
+transactions lead to zero inflation. This sounds worse than it is, since locally, every individual transaction is
+checked for validity at the time of inclusion in the blockchain.
+
+We presume however, that the intervening transactions were all valid, since they all passed consensus at the time of
+inclusion.
+
+This is why we can discard range proofs after UTXOs are spent. We presume that the UTXO chain that led to the current
+set were all checked and had positive values, but we only really care that the current UTXO set contains all
+non-negative values.
+
+This argument is independent of the nature of the additional restrictions. Specifically, if these restrictions are
+manifested as a script that provides additional constraints over whether a UTXO may be spent, the same arguments apply:
+We assume that the rules were properly enforced up until now, but it's not really our concern. What we care about is
+that no inflation has occurred to this point and that we enforce all the consensus rules for any new transactions going
+forward.
+
+This means that from a philosophical viewpoint, there ought to be no reason that Tari Script is not workable, and
+further, that pruning spent outputs (and possibly the scripts associated with them) is no different from pruning range
+proofs.
+
+### Additional requirements
+
+The assumptions that broadly equate scripting with range proofs in the above argument are:
+
+* The script (hash) must be committed to in the blockchain.
+* The script must not be malleable in any way without invalidating the transaction.
+* The creator of the UTXO must commit to and sign the script (hash).
```

Comment:
Cant we just incorporate it in the challange?
That should ensure its not malleable, is signed and is commit in the blockchain. 
I made a case we I think its better to put this in the kernel

Reply:
Same issue. It's bad for privacy.

Reply:
Agreed.